Allow parsing of overflow-anchor behind flag

nmoucht opened the pull request:
#### 397d2db0121ca1dc5b39ab523c85586cedba4710
<pre>
Allow parsing of overflow-anchor behind flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=242973">https://bugs.webkit.org/show_bug.cgi?id=242973</a>
&lt;rdar://97694545&gt;

Reviewed by NOBODY (OOPS!).

Add flag for scroll anchoring and allow parsing of overflow-anchor behind
that flag.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/parsing/overflow-anchor-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/parsing/overflow-anchor-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/cssom/getComputedStyle-detached-subtree-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::operator OverflowAnchor const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeOverflowAnchor):
(WebCore::CSSPropertyParser::parseSingleValue):
* Source/WebCore/platform/ScrollTypes.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::overflowAnchor const):
(WebCore::RenderStyle::setOverflowAnchor):
(WebCore::RenderStyle::initialOverflowAnchor):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
</pre>

smfr commented on Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
```diff
@@ -405,6 +405,18 @@ CSSRelativeColorSyntaxEnabled:
+CSSScrollAnchoringEnabled:
+  type: bool
+  humanReadableName: "CSS Scroll Anchoring"
+  humanReadableDescription: "Enable CSS Scroll Anchoring support"
```
The word "support" doesn't add anything.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -1907,6 +1907,37 @@ template<> inline CSSPrimitiveValue::operator OverscrollBehavior() const
     return OverscrollBehavior::Auto;
 }
 
+template<> inline CSSPrimitiveValue::CSSPrimitiveValue(OverflowAnchor a)
```
Please use a longer name than `a`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
```diff
@@ -252,6 +252,7 @@ class StyleRareNonInheritedData : public RefCounted<StyleRareNonInheritedData> {
     LengthPoint offsetPosition;
     LengthPoint offsetAnchor;
     OffsetRotation offsetRotate;
+    unsigned overflowAnchor : 1; // Scroll Anchoring- Overflow Anchor
```
More this up next to the other bitfield entries to pack things better.

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -7966,6 +7976,12 @@
+            "longname": "CSS Scroll Anchoring Module",
+            "url": "https://www.w3.org/TR/css-scroll-anchoring-1/"
+        },
+
```
nit: superfluous blank line

nmoucht replied:
Fixed.

nt1m reviewed (changes requested):
https://github.com/WebKit/WebKit/commit/d70cd13f28142083ebffc879426101ad5a75e0d9 changed the way CSS property settings work, you may want to follow examples in there.

nt1m commented on Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
```diff
@@ -243,6 +243,8 @@ class StyleRareNonInheritedData : public RefCounted<StyleRareNonInheritedData> {
 
     unsigned containerType : 2; // ContainerType
 
+    unsigned overflowAnchor : 1; // Scroll Anchoring- Overflow Anchor
```
```suggestion
    unsigned overflowAnchor : 1; // Scroll Anchoring - OverflowAnchor
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -3832,6 +3832,16 @@
                 "url": "https://www.w3.org/TR/css-overflow-3/#propdef-overflow"
             }
         },
+        "overflow-anchor": {
```
```suggestion
        "overflow-anchor": {
            "codegen-properties": {
                "settings-flag": "cssScrollAnchoringEnabled"
            },
```

This is how you link the property to the setting.

nmoucht replied:
Ah I see, thanks.

nt1m commented on Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
```diff
@@ -4126,7 +4126,10 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
+        case CSSPropertyOverflowAnchor:
+            if (!m_element->document().settings().cssScrollAnchoringEnabled())
+                return nullptr;
+            return cssValuePool.createValue(style.overflowAnchor());
```
```suggestion
        case CSSPropertyOverflowAnchor:
            return cssValuePool.createValue(style.overflowAnchor());
```

Since the property will never be parsed with the flag disabled, the style will never be computed in that case.

nmoucht replied:
Fixed.

nt1m reviewed (approved):
LGTM assuming you've manually tested the preference works (the new mechanism is new to me as well, so I'm not 100% sure)

nmoucht committed:
Allow parsing of overflow-anchor behind flag
https://bugs.webkit.org/show_bug.cgi?id=242973

Reviewed by Simon Fraser and Tim Nguyen.

Add flag for scroll anchoring and allow parsing of overflow-anchor behind
that flag.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::CSSPrimitiveValue::CSSPrimitiveValue):
(WebCore::CSSPrimitiveValue::operator OverflowAnchor const):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::operator==):
(WebCore::add):
(WebCore::CSSParserContext::isPropertyRuntimeDisabled const):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeOverflowAnchor):
(WebCore::CSSPropertyParser::parseSingleValue):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::overflowAnchor const):
(WebCore::RenderStyle::setOverflowAnchor):
(WebCore::RenderStyle::initialOverflowAnchor):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:

Canonical link: https://commits.webkit.org/252906@main
